WIP: Ask to uninstall if already installed

fredrikaverpil opened the pull request:
I need to make sure users are uninstalling the app prior to installing a new version. I'm quite unfamiliar with NSIS and have managed to make a new window pop up on top of the installer wizard, which will ask the user to uninstall if the app is already installed. This is a very crude way to go about this, and I actually would've preferred if this step was part of the actual wizard rather than a window on top of the wizard.

I was a bit afraid that the uninstall window could appear below the regular wizard window, and I just recently received a report from a user that they mistakenly installed the app and then saw the uninstall window... so it might be possible that this could actually happen...

I guess this PR is just a starting point and something we can use to discuss this feature, as I don't feel this is polished enough to merge yet.

I got the code from here: http://nsis.sourceforge.net/Auto-uninstall_old_before_installing_new

mherrmann commented:
Thank you for the PR Fredrik and sorry for getting back to you so late. The uninstall window appearing underneath the regular window is indeed a problem. Maybe [this](https://github.com/kichik/nsis/blob/1de948af9683ccd04add5fe5e4cea35ecc576d4b/Examples/makensis.nsi#L907) could serve as an alternative solution?

fredrikaverpil commented:
I ended up reading the docs a bit and have now created a shared function which is executed from both the installer and the uninstaller. This means that in the installation window where you can see what is being copied into the `$InstDir`, you first see a bunch of `Removing...` entries as the `RMDir /r "$InstDir"` command is executing.

This seems nicer, at least for my usage.

I've also come to realize a couple of things:

* If the user installs into `C:\Program Files` or similar bad paths, the uninstaller will attempt to perform the `RMDir /r` on this path, which would effectively break a lot of things... I haven't implemented anything to prevent this from happening but thought this was worth mentioning here.
* If the `RMDir` cannot remove files because they are e.g. locked, you cannot really query which files were not deleted or easily indicate this to the user. I opted to run a custom function after the `RMDir` command has completed. This custom function checks whether `$InstDir` still exists and if so, it also checks if the directory is empty. If not, it will show a message box. All in all, this is a very crude implementation, which basically results in a dialog popping up, asking the user to empty the folder manually...

I've also added the ability to launch the application at the end of a silent installation by reading a custom commandline argument.

So this brings me to... of all this, what could possibly be interesting to contribute back to fbs? I am happy to share my code. What do you think @mherrmann ?

mherrmann commented:
Thank you very much for your further thoughts and offer to contribute Fredrik. It could be great to incorporate your "real-world" learnings.

One thing I'm curious about: You said

> I need to make sure users are uninstalling the app prior to installing a new version.

Why is that the case? Wouldn't installing a new version simply overwrite the existing files?

The solution you have sounds cool. So if I understand correctly, the app simply silently uninstalls itself when you re-install it? (Silently as in it does not prompt the user beforehand. As you mentioned, it displays the files that are being deleted.)

You're right `RMDir /r` is dangerous. Maybe the uninstaller should only remove files it actually installed? Eg. using [this macro](http://nsis.sourceforge.net/Uninstall_only_installed_files).

From my current limited understanding, I think a ideal solution would have the following properties:

 * The Uninstaller only removes files which the Installer actually created. (No more `RMDir /r`.)
 * If the Installer is run again when the app is already installed, then it prompts "App is already installed. Do you want to re-install it?".
 * If the installation directory is non-empty when the Uninstaller completes, it shows a message: "Some files could not be deleted. You may have to delete C:\Program Files\App manually.".

What do you think?

Regarding the command line flag for launching the app from the silent installer, I don't currently think that would be relevant for fbs.

Cheers!
Michael

fredrikaverpil commented:
> One thing I'm curious about: You said
>>    I need to make sure users are uninstalling the app prior to installing a new version.
>
> Why is that the case? Wouldn't installing a new version simply overwrite the existing files?

Yes. And if there are file lockings or permission issues, the installer will ask the user to ignore/retry/abort. But what I'm concerned about is when old files from previous installations are left in the installation, without being removed. For example, I switched my app's Python from 3.6 to 3.7 a while ago. This would've left `python36.dll` in there as well as possibly other files which may have been deprecated or whatever... which could result in unexpected results. This is why I want to completely erase the install folder on each reinstall.

> The solution you have sounds cool. So if I understand correctly, the app simply silently uninstalls itself when you re-install it? (Silently as in it does not prompt the user beforehand. As you mentioned, it displays the files that are being deleted.)

Exactly. I have reverted my previous commit and updated the PR with the shared uninstall function which just has the same logic now as your original uninstaller. You can run this and see for yourself how this works. It's pretty slick.

During the uninstall, the `RMDir` will not complain when a file cannot be deleted. The NSIS script just silently continues. This is for me undesired behavior, and I have implemented [isEmptyDir](http://nsis.sourceforge.net/Check_if_dir_is_empty) which will warn the user via a simple dialog box when such issues are hit. This is not a polished solution as it does not roll back anything or offer the ability to retry etc. It's just a warning where I actually ask the user to manually delete the install folder and completely re-install. Since this is for an inhouse application, it's fine for our use case, but would be a very crude implementation for a more general use case.

> I think a ideal solution would have the following properties:
>- The Uninstaller only removes files which the Installer actually created. (No more RMDir /r.)
>- If the Installer is run again when the app is already installed, then it prompts "App is already installed. Do you want to re-install it?".
>- If the installation directory is non-empty when the Uninstaller completes, it shows a message: "Some files could not be deleted. You may have to delete C:\Program Files\App manually.".

Having the uninstaller only remove the files which were actually created would be much, much better.

I might want to skip the actual question on reinstalling, and just simply reinstall. But for general use cases, your suggestion with showing the question is more informative.

When the directory is non-empty, the question is whether to continue with the installation or not...?

mherrmann commented:
> This would've left python36.dll in there as well as possibly other files which may have been deprecated or whatever... which could result in unexpected results. This is why I want to completely erase the install folder on each reinstall.

Makes sense.

> I have reverted my previous commit and updated the PR

Thanks. It's an elegant starting point. In my eyes, it still suffers from a few limitations though:

 * It silently `RMDir`s the previous installation directory. If there are any files in there that need to be kept, they will disappear - without the user having been warned. (I know you're aware of this.)
 * Instead of executing the uninstaller logic when re-installing itself, it should ideally use the registry to locate the uninstaller of the installed version, and invoke that. The reason is that the uninstallation logic may change between version. If that happens, we wouldn't want to run the uninstall logic of v2 to remove v1.

> When the directory is non-empty, the question is whether to continue with the installation or not...?

I think I would continue. Removing the previous installation is only a safety measure anyways.

It seems to me that re-installing is quite a bit more subtle than it appears at first :flushed:

fredrikaverpil commented:
> Instead of executing the uninstaller logic when re-installing itself, it should ideally use the registry to locate the uninstaller of the installed version, and invoke that. 

Yes, I kind of agree with you. But the problem I wanted to solve initially was that this invocation can actually cause the window to open up below the installer window...
